File: Algorithms/matrix_power/matrix_power.cpp


#define MAX_N 2  //example for matrix fibonnaci numbers


struct Matrix { int mat[MAX_N[MAX_N]; };

Matrix matrixMultiplication(Matrix A, Matrix B) {
	Matrix ans; 
	for(int i=0; i<MAX_N; ++i) 
		for(int j=0; j<MAX_N; ++j) 
			for(int k=0; k<MAX_N; ++k) 
				ans.mat[i][j] += A.mat[i][k] * B.mat[k][j];
	return ans;
}

Matrix matrixPower(Matrix Base, int p) {
	Matrix ans;
	for(int i=0; i< MAX_N; ++i)
		for(int j=0; j< MAX_N; ++j)
			ans.mat[i][j] = (i==j);
	while(p) {
		if(p&1)
			ans = matrixMultiplication(ans, Base);
		Base = matrixMultiplication(Base, Base);
		p >>=1;
	}
	return ans;
}

File: Algorithms/matrix_power/matrix_power.py
MAX_N = 2 #change for larger matrices

def multiply(A, B):
	ans = [[0 for i in range(MAX_N)] for j in range(MAX_N)] 
	for i in range(MAX_N): 
		for j in range(MAX_N): 
			for k in range(MAX_N):
				ans.mat[i][j] += A[i][k] * B[k][j];
	return ans;


def power(F, n):
    if n <= 1:
        return
    power(F, n // 2)
    multiply(F, F)
    if n % 2 != 0:
        F = multiply(F, F)
    



